Add spec for StartInBackgroundMode.md (#2781)
MicrosoftEdge/WebView2Feedback: specs/StartInBackgroundMode.md

/// `ICoreWebView2EnvironmentOptions3:CreationPriority` option.
[v1_enum]
typedef enum COREWEBVIEW2_CREATION_PRIORITY {
  COREWEBVIEW2_CREATION_PRIORITY_NORMAL,
  COREWEBVIEW2_CREATION_PRIORITY_HIGH,
} COREWEBVIEW2_CREATION_PRIORITY;

[uuid(efb58776-32bd-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2EnvironmentOptions3 : ICoreWebView2EnvironmentOptions2 {

  /// Gets the creation priority for webview2 creation.
  ///
  /// \snippet AppWindow.cpp CreateCoreWebView2EnvironmentWithOptions
  // MSOWNERS: wangsongjin@microsoft.com
  [propget] HRESULT CreationPriority([out, retval] COREWEBVIEW2_CREATION_PRIORITY* creationPriority);

  /// Sets the `CreationPriority` property.
  /// The `CreationPriority` property specifies the priority used during the WebView2 
  /// browser process creation.
  /// Set `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_NORMAL` will create the
  /// WebView2 browser process with normal priority.
  /// Set `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_HIGH` will attempt to
  /// create the WebView2 browser process with high priority.
  /// Default is `COREWEBVIEW2_CREATION_PRIORITY_HIGH`.

>   /// Note that 1)The host app's priority must be at least normal for 

Reviewer: ```suggestion
  /// Note that the host app's priority must be at least normal for 

```

>   /// `COREWEBVIEW2_CREATION_PRIORITY_HIGH` to be applied. Else, the WebView2 browser process 

Reviewer: ```suggestion
  /// `COREWEBVIEW2_CREATION_PRIORITY_HIGH` to be applied. Otherwise the WebView2 browser process 

```

  /// still get created with normal priority even with `COREWEBVIEW2_CREATION_PRIORITY_HIGH` setting.

>   /// 2)Currently `CreationPriority` only applies to the creation of the WebView2 browser process.

Reviewer: ```suggestion
  /// Additionally, currently `CreationPriority` only applies to the creation of the WebView2 browser process.

```

CreationPriority
===

# Background

> WebView2 has a functionality enabled by default which raises the launch priority of WV2

Reviewer: ```suggestion
WebView2 has functionality enabled by default which raises the launch priority of WebView2
```

> processes to ensure that WV2 processes are given higher CPU priority and access to resources 

Reviewer: ```suggestion
processes to ensure that WebView2 processes are given higher CPU priority and access to resources 
```

> while creation, and that can bring improvements on time spent creating a WV2 instance. Since 

Reviewer: ```suggestion
during creation, and that can bring improvements on time spent creating a WebView2 instance. Since 
```

> this improvement is at the cost of taking higher CPU priority, we introduce `CreationPriority` 

Reviewer: ```suggestion
this improvement is at the cost of taking higher CPU priority, we introduce the `CreationPriority` 
```

property to give developers an option to decide by themselves.

# Description
* Setting `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_NORMAL` will signal WebView2 

> to use less system resources during the creation. 

Reviewer: Is this correct? Its the creation of the browser process that benefits from the priority?

```suggestion
to have normal priority compared to other processes during the creation of the WebView2 browser process.
```

Author: "use less system resources" is not accurate. Rephased!

* Setting `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_HIGH` will attempt to enable 

> the launch time improvement functionality with raising the launch priority during the creation.  

Reviewer: ```suggestion
raise the priority of the creation of the WebView2 browser process.
```

* Default value is `COREWEBVIEW2_CREATION_PRIORITY_HIGH`.

> ## Notes

Reviewer: These notes seem relevant to our reference documentation. They should be included in the reference documentation of the CreationPriority property below.

> * The host app's priority is at least normal before applying the launch time improvement functionality. 

Reviewer: I would phrase this bullet as more like:

```
Applying high priority to WebView2 creation is best effort. High priority is unavailable on ... {when is high priority not available - on previous versions of Windows?}. When high priority is not available, normal priority will be used instead.
```

Else, WebView2 still get created with normal priority even with `COREWEBVIEW2_CREATION_PRIORITY_HIGH` 
setting.

> * Currently launch time improvement functionality is only enabled for WebView2 manager process, so the 

Reviewer: We don't document the launch time improvement feature and we don't need to talk about it in our reference documentation. We can just document the CreationPriority feature directly.

```suggestion
* Currently
```

> `CreationPriority` only effects the creation of manager process. `CreationPriority` can play the same 

Reviewer: ```suggestion
`CreationPriority` only applies to the creation of the WebView2 browser process. `CreationPriority` may be broadened to
```

> role for other process types like render process if WebView2 opts in launch time improvement functionality 

Reviewer: ```suggestion
apply to other parts of WebView2 creation
```

> for those types in the future.

Reviewer: ```suggestion
in the future.
```

# Examples
## Win32 C++
```cpp
auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();
// Set `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_NORMAL` if you don't want to boost the 
// launch time at the cost of using higher CPU priority during the launch phase.

> options->put_CreationPriority(COREWEBVIEW2_CREATION_PRIORITY_NORMAL);

Reviewer: Why would an end developer want to have normal pri for their WebView2 creation? Can we work the answer to that question into the sample code? That is, can the sample code show why we're setting the creation priority? (If not, perhaps just adding a comment about **why** we're setting normal pri in the sample code)

Author: > Why would an end developer want to have normal pri for their WebView2 creation?

The use case for normal pri is end developer does not want to receive the higher CPU pri.

>  That is, can the sample code show why we're setting the creation priority?

No, I don't think so. 

> If not, perhaps just adding a comment about why we're setting normal pri in the sample code

What about "Set `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_NORMAL` to prevent applying the higher CPU priority during WebView2 browser process Creation".?

Reviewer: > The use case for normal pri is end developer does not want to receive the higher CPU pri.

But why don't they want higher CPU priority?

     this, &AppWindow::OnCreateEnvironmentCompleted).Get());
```
## WinRT and .NET
```c#
CoreWebView2EnvironmentOptions options = new CoreWebView2EnvironmentOptions();
options.CreationPriority = COREWEBVIEW2_CREATION_PRIORITY_NORMAL;
CoreWebView2Environment Environment = 
    await CoreWebView2Environment.CreateAsync(
        BrowserExecutableFolder, UserDataFolder, options)
```

# API Details
## Win32 C++
```cpp
/// Specifies the type of webview2 creation priority for the
/// `ICoreWebView2EnvironmentOptions3:CreationPriority` option.
[v1_enum]
typedef enum COREWEBVIEW2_CREATION_PRIORITY {
  COREWEBVIEW2_CREATION_PRIORITY_NORMAL,
  COREWEBVIEW2_CREATION_PRIORITY_HIGH,
} COREWEBVIEW2_CREATION_PRIORITY;

[uuid(efb58776-32bd-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2EnvironmentOptions3 : ICoreWebView2EnvironmentOptions2 {

  /// Gets the creation priority for webview2 creation.
  ///
  /// \snippet AppWindow.cpp CreateCoreWebView2EnvironmentWithOptions
  // MSOWNERS: wangsongjin@microsoft.com
  [propget] HRESULT CreationPriority([out, retval] COREWEBVIEW2_CREATION_PRIORITY* creationPriority);

  /// Sets the `CreationPriority` property.

>   /// The `CreationPriority` property specifies that the creation priority WebView

Reviewer: ```suggestion
  /// The `CreationPriority` property specifies the priority used during WebView2 browser process creation.

```

  /// environment attempt to create.
  /// Set `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_NORMAL` will create

>   /// WebView2 with normal priority.

Reviewer: ```suggestion
  /// the WebView2 browser process with normal priority.

```

  /// Set `CreationPriority` to `COREWEBVIEW2_CREATION_PRIORITY_HIGH` will attempt

>   /// to create WebView2 with high priority.

Reviewer: ```suggestion
  /// to create the WebView2 browser process with high priority.
```

StartInBackgroundMode
===

# Background

> By default, WebView2 instances get created with some priorities set higher, assuming the WV2 is being created in the foreground for an important purpose. However, not every developer wants to boost the WebView2 prelaunch phase at the cost of utilizing more system resources. The `StartInBackgroundMode` property specifies that whether WebView2 created with launch boost or not.

Reviewer 2: Please don't use the word "boost". It has a pretty specific meaning to many people and that's not what's happening in our case.

Author: Will update!

# Description
Setting `StartInBackgroundMode` to TRUE will signal WebView2 to start in a 'background mode' which will attempt to use less system resources during the creation and initial load. Default is FALSE.

# Examples
## Win32 C++
```cpp
auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();
options->put_StartInBackgroundMode(TRUE));
HRESULT hr = CreateCoreWebView2EnvironmentWithOptions(
  nullptr, m_userDataFolder.c_str(), options.Get(),
  Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>(
     this, &AppWindow::OnCreateEnvironmentCompleted).Get());
```
## WinRT and .NET
```c#
auto options = new CoreWebView2EnvironmentOptions();
options.StartInBackgroundMode = true;
auto environment = await CoreWebView2Environment.CreateAsync(BrowserExecutableFolder, UserDataFolder, options);
```

# API Details
## Win32 C++
```cpp

> interface ICoreWebView2StagingEnvironmentOptions3 : IUnknown 

Reviewer: We should use the expected public API names in the API spec so exclude the 'Staging' and derive from the expected parent (ICoreWebView2EnvironmentOptions2?)

Author: Yep. Will use ICoreWebView2EnvironmentOptions3 instead.

{
  /// Setting `StartInBackgroundMode` to `TRUE` will signal WebView2 to start in a 'background mode'

>   /// which will attempt to use less system resources during the creation and initial load.

Reviewer: 'Creation' and 'initial load' are ambiguous - we're not saying creation of what and load is also ambiguous maybe talking about the first navigation? The documentation should make it clear when this behavior will take place in terms of WebView2 APIs or other documented WebView2 things. Is this low resource behavior during each CoreWebView2 creation and ending when the async method returns the controller to the caller? Or is it maybe only when starting the browser process?

Author: This will function on each WV2 process creation not limited on manager/browser process only. 
> 'Creation' and 'initial load' are ambiguous

Yep, I'll rephase it!

Author: I should make it more clear here. This new property works for disabling launch boost for all processes. For launch boost, we currently only use it for Manager process,  child processes like Utilities and Render is only for experimental purpose (under sandbox). 

It's a good call. I should clarify that on the doc.

  /// Default is FALSE.

>   [propget] HRESULT StartInBackgroundMode([out, retval] BOOL* value);

Reviewer: Properties should be named with noun phrases whereas Methods and Events use verb phrases. This is named 'Start...' makes it sound like a method. We could name this 'Should...' prefix instead.

Author: Make sense. Will update!

>   [propget] HRESULT StartInBackgroundMode([out, retval] BOOL* value);

Reviewer: Saying 'start in background mode' makes it sound like WebView2 has entered this mode and its not clear how or when it will exit. So rather than 'start in .. mode' I'd rather something that ties it specifically to when this low priority behavior will apply. Like ShouldCreateWithLowPriority or ShouldCreateWithMinimalResourceUsage or something?

Reviewer: Also the term 'background' is a bit ambiguous as well. It has some ties to UI when this feature is really about resource usage / priority.

Author: Yep. I agree we should use something like `ShouldCreateWithLowPriority` or `ShouldCreateWithMinimalResourceUsage` 

From the [reference](https://microsoft.sharepoint.com/:w:/r/teams/Edge/_layouts/15/Doc.aspx?sourcedoc=%7B8C776052-C93A-4EE1-8658-2D8182757B19%7D&file=Launch%20Priority%20Announcement%20-%20August%202021.docx&action=default&mobileredirect=true&share=IQFSYHeMOsnhToZYLYGCdXsZAcFBNjNDc_-s7rGB6eXHRaw&cid=7a8fc950-9a02-45a5-a8b4-7f1d6a78038c&wdLOR=c4DF38799-7315-4F39-87AF-F4ABB139DFAD)
> When an application using WebView2 is launching, there’s a period of time before foreground boost kicks-in and offers benefits. During this period WV2 processes are susceptible to starvation. We’re adding WV2 functionality that will ship in Edge 94 that __raises the launch priority of WV2 processes to ensure that WV2 processes are given higher CPU priority and access to resources.__ This will give the manager process higher CPU priority until the application has a foreground process that will trigger foreground boost. 

Setting this property to `TRUE` just disables `launch_boost` which actively raises the priority. The code inside [launch_win.cc](https://microsoft.visualstudio.com/Edge/_git/chromium.src?path=%2Fbase%2Fprocess%2Flaunch_win.cc&_a=contents&version=GBmain)(Line 442). 

So i'm not able to make choice around `ShouldCreateWithLowPriority` or something like  `ShouldCreateWithoutRaisingPriority`.

Reviewer: I would bet foreground boost won't exist on other OSes. Maybe we can make this an enum property instead and name it CreationPriority = Default/High/Normal and say in the docs this is best effort

Author: > I would bet foreground boost won't exist on other OSes. 

Yep. Only Win. Good point! I will add description for this.

> Maybe we can make this an enum property instead and name it CreationPriority = Default/High/Normal and say in the docs this is best effort

Good alternative! 
__Default, High__ represents create with trying to Enable launch boost.
  
One thing bothers me is this property is not the only signal determines if we create WV2 with higher Priority. `CanLaunchManagerProcessWithBoost()` also contribute. 
```c++ 
bool use_launch_boost =
      CanLaunchManagerProcessWithBoost() && !(creationPriority == CreationPriority.Normal);
TRACE_EVENT_INSTANT1("edge_webview", "WebView2_ManagerProcessLaunchBoost",
                       TRACE_EVENT_SCOPE_GLOBAL, "enabled", use_launch_boost);
```
> `CanLaunchManagerProcessWithBoost()` function defines we want to assure that we are only providing startup boost on desktop, and that the host app's priority is at least normal. [Code Link](https://microsoft.visualstudio.com/Edge/_git/chromium.src?path=%2Fedge_embedded_browser%2Fclient%2Fwin%2Futil.cc&_a=contents&version=GBmain) Line 587

So `CreationPriority` for me is more like **"Trying to"**, we can say that in doc. 

Well, I just think about another naming `ShouldCreateWithSavinglResourceUsage` which directly shows the benefit of why setting this property.

Do you have strong preference between`ShouldCreateWithSavinglResourceUsage` and enum `CreationPriority`? I tend to use `ShouldCreateWithSavinglResourceUsage`

Author: Offline confirmed with Dave. The reason we choose to use enum instead of `ShouldCreateWithLowPriority` are: 
- We're not really "low priority" with applying this feature and we're like normal priority. `ShouldCreateWithLowPriority` is not accurate.
-  For  `ShouldCreateWithLowPriority`, if we want to have a real low priority feature in the future we won't know the difference. The enum give us a chance to add Low if we want to support that.

Reviewer: I'm OK with ShouldCreateWithLowPriority only if we know we'll never want to make an actual low pri feature. Otherwise this is painting us into a corner

  /// Sets the `StartInBackgroundMode` property.
  [propput] HRESULT StartInBackgroundMode ([in] BOOL value);
}
```
## WinRT and .NET
```c#

> unsealed runtimeclass CoreWebView2EnvironmentOptions

Reviewer: You should have the namespace declaration around this as well. See https://github.com/MicrosoftEdge/WebView2Feedback/pull/2779/files for example

Author: Sure. Will add!

CreationPriority
===

# Background

> WebView2 has a functionality enabled by default which raises the launch priority of WV2 processes to ensure that WV2 processes are given higher CPU priority and access to resources while creation, and that can bring improvements on time spent creating a WV2 instance. Since this improvement is at the cost of taking higher CPU priority, we introduce `CreationPriority` property to give developers an option to decide by themselves.

Author: Noted!

# Description
* Setting `CreationPriority` to `CreationPriority::NORMAL` will signal WebView2 to use less system resources during the creation. 
* Setting `CreationPriority` to `CreationPriority::HIGH` will attempt to enable the launch time improvement functionality with raising the launch priority during creation.  
* `CreationPriority::DEFAULT` is the default value, we treat it as same as `CreationPriority::HIGH`.

> ## Notes

Author: Agree! Will move it to the reference documentation.

> * Your application needs to verify that the version of Edge is >=95 before applying the `CreationPriority`, it can cause issues if used on earlier versions of Edge.

Author: Yep, set to `normal` should not be applied to older version, I got this from Launch priority Announcement doc. The existing is background flag bring this limitation. 

>  Can we detect this case in our code and not apply the value in that case?
Agree!

> * Your application needs to verify that the version of Edge is >=95 before applying the `CreationPriority`, it can cause issues if used on earlier versions of Edge.

Author: > Can we detect this case in our code and not apply the value in that case?

This is fine as we won't apply this new property is any version < 95. This new API with shipped with currently version. 
I should remove this.

* A prerequisitive for WebView2 applying the launch time improvement is the host app's priority is at least normal. Else, WebView2 still get created with normal priority even with `CreationPriority::HIGH`.
* Currently launch time improvement functionality is only enabled for WebView2 manager process, so the `CreationPriority` only effect the creation of manager process. `CreationPriority` can play the same role for other process types like render process if WebView2 opts in launch time improvement functionality for those types in the future.
# Examples
## Win32 C++
```cpp
auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();

> options->put_CreationPriority(CreationPriority::NORMAL);

Author: I can add a description for use case.
>  Is there sample code we can include here that can make it clear what the end to end scenario is rather than just showing setting the property and calling Create?

`CreationPriority` is only used in creation WebView environment, I don't think there're more reference codes here.

HRESULT hr = CreateCoreWebView2EnvironmentWithOptions(
  nullptr, m_userDataFolder.c_str(), options.Get(),
  Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>(
     this, &AppWindow::OnCreateEnvironmentCompleted).Get());
```
## WinRT and .NET
```c#

> auto options = new CoreWebView2EnvironmentOptions();

Author: Oh yes, falsely copied from WinRT. Refined!

options.CreationPriority = CreationPriority::NORMAL;
auto environment = await CoreWebView2Environment.CreateAsync(BrowserExecutableFolder, UserDataFolder, options);
```

# API Details
## Win32 C++
```cpp
[uuid(efb58776-32bd-11ed-a261-0242ac120002), object, pointer_default(unique)]
interface ICoreWebView2EnvironmentOptions3 : ICoreWebView2EnvironmentOptions2 {

  /// Gets the creation priority for webview2 creation.
  // MSOWNERS: wangsongjin@microsoft.com

>   [propget] HRESULT CreationPriority([out, retval] CreationPriority* creationPriority);

Author: Noted!

  /// Sets the `CreationPriority` property.
  /// The `CreationPriority` property` specifies that the creation priority WebView environment get created.

>   /// WebView2 currernt has a functionality enabled by deafult which raises the launch priority to

Author: Make sense! Moved it the description.

  /// give improvements on the WebView2 creation time. Since this improvement is at the cost of occupying  
  /// higher CPU priority, we give developers options to decide by themselves.
  /// Set `CreationPriority` to `CreationPriority::NORMAL` will create WebView2 instances with normal priority 
  /// which opts out this functionality. 

>   /// Set `CreationPriority` to `CreationPriority::HIGH` will try to launch WebView2 with above normal priority  

Author: Update it to high.

  /// and opt in this functionality.
  /// Set `CreationPriority` to `CreationPriority::DEFAULT` behaves the same way as `CreationPriority::HIGH`. 
  /// NOTE: A prerequisitive for WebView2 applying the launch time improvement is the host app's priority is
  /// at least normal. Else, WebView2 still get created with normal priority even with `CreationPriority::HIGH`.
  // MSOWNERS: wangsongjin@microsoft.com
  [propput] HRESULT CreationPriority([in] CreationPriority creationPriority);
}
```
## WinRT and .NET
```c#
namespace Microsoft.Web.WebView2.Core
{
    // ...
    unsealed runtimeclass CoreWebView2EnvironmentOptions
    {

>         enum CreationPriority { get; set; };

Author: Gotcha! Will run the script.
